beps/0005-definitive-plugin-listing/README.md

```diff
@@ -0,0 +1,250 @@
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs or even code snippets.
+If there's any ambiguity about HOW your proposal will be implemented, this is the place to discuss them.
+-->
+
+### Glossary
+
+### New Type `BackstageFeature`
+
+```ts
+type BackstageFeature =
+  | {
+      type: 'plugin';
+      pluginId: string;
+    }
+  | {
+      type: 'module';
+      pluginId: string;
+      moduleId: string;
+    };
+```
+
+### New `InstanceMetadataService`
+
+While we could attach the existing information to the `PluginMetadataService`, we propose a new service that handles instance-level information. The existing `PluginMetadataService` should reveal information about the plugin itself, its `pluginId`, dependencies or similar. The new `InstanceMetadataService` should give you information about the entire Backstage instance that you're interrogating. One could imagine this service also having information about instance URLs, health or gateway status.
+
+```ts
+interface InstanceMetadataService {
+  getInstalledFeatures: () => BackstageFeature[];
+}
+```
+
+Currently, this would need to be bootstrapped directly in `BackendInitializer` or `BackstageBackend`. It may make sense to create a new `FeatureRegistry` class that can expose this instead similar to the `PluginMetadataService`.
+
+### New `DeploymentMetadataService`
+
+This will be a new service that can be provided either as a core service or a new package. The idea is that it will show all information across your entire deployment. One can imagine this hosting aggregate information around instances in the deployment, like their internal and external URLs.
+
+```ts
+interface DeploymentMetadataService {
+  getInstalledFeatures: () => BackstageFeature[];
+}
+```
+
+By default, this will either
+a. be the exact same as `InstanceMetadataService` if we are not using `HostDiscovery` static config, or
+b. this will be pulled from that `HostDiscovery` static config.
+
+That would look something like,
+
+```yaml
+discovery:
+  endpoints:
+    - target: https://internal.example.com/internal-catalog
+      plugins: [catalog]
+    - target: https://internal.example.com/secure/api/{{pluginId}}
+      plugins: [auth, permission]
+    - target:
+        internal: https://internal.example.com/search
+        external: https://example.com/search
+      plugins: [search]
+```
+
+So, the above interface would return for the above config,
+
+```ts
+const installedFeatures = deploymentMetadataService.getInstalledFeatures();
+// installedFeatures = [{type: 'plugin', pluginId: 'catalog'}, {type: 'plugin', pluginId: 'auth'}, {type: 'plugin', pluginId: 'permission'}, {type: 'plugin', pluginId: 'search'}]
```

Comment:
As is, this implementation would only return plugins and nothing dynamic since it isn't calling into the individual instance. I'll call that out better here, I was trying to provide a solid static implementation that was extendable once we get more dynamic.

I was trying to scope this to a static implementation for ease of use and less config/routing problems. Happy to go down the route of URL based discovery, but I was thinking that would be a v2 of this BEP and not immediately implemented.

Reply:
Coming back to this, this is a poor example that matches HostDiscovery but doesn't demonstrate really what I was trying to do here. `moduleId` is difficult to do without the dynamic discovery this BEP had originally. We could update this to URL based, ie initially using the list of URLs configured in the discovery config and working backwards based on their instance metadata endpoints - I would like that to be a v2 of this proposal, it definitely makes sense, but has more moving parts that I think implementing the static version of would help flush out.

Reply:
Coming back to this again - I think URLs is the right answer here. It solves the issue with modules not showing up in my initial design and ensures that data will be consistent if configs are different across machines. I'll update the doc with that.